lib/opte/src/engine/port.rs

```diff
@@ -1460,6 +1483,147 @@ impl<N: NetworkImpl> Port<N> {
         }
     }
 
+    /// Creates a new TCP flow state entry for a given packet.
+    ///
+    /// # Errors
+    /// * `OpteError::MaxCapacity(_)` if the TCP flow stable is full.
+    /// * `ProcessError::TcpFlow(_)` if we do not have a valid transition from
+    ///   `Closed` based on the packet state.
+    fn create_new_tcp_entry(
+        &self,
+        tcp_flows: &mut FlowTable<TcpFlowEntryState>,
+        tcp: &TcpMeta,
+        dir: &TcpDirection,
+        pkt_len: u64,
+    ) -> result::Result<TcpState, ProcessError> {
+        // Create a new entry and find its current state. In
+        // this case it should always be `SynSent`, unless we're
+        // recovering an `Established` flow.
+        let mut tfs = TcpFlowState::new();
+
+        let tcp_state = match tfs.process(
+            self.name_cstr.as_c_str(),
+            dir.dir(),
+            dir.local_flow(),
+            tcp,
+        ) {
+            Ok(tcp_state) => tcp_state,
+
+            // We're intentionally not allowing through unexpected segments or
+            // new flow conditions: SYN packets will always be accepted in the
+            // starting states, and we have valid shortcuts back into `Established`.
+            Err(e) => return Err(ProcessError::TcpFlow(e)),
+        };
+
+        if tcp_state != TcpState::Closed {
+            // The inbound UFID is determined on the inbound side.
+            let (ufid_out, tfes) = match *dir {
+                TcpDirection::In { ufid_in, ufid_out } => (
+                    ufid_out,
+                    TcpFlowEntryState::new_inbound(*ufid_in, tfs, pkt_len),
+                ),
+                TcpDirection::Out { ufid_out } => {
+                    (ufid_out, TcpFlowEntryState::new_outbound(tfs, pkt_len))
+                }
+            };
+            match tcp_flows.add(*ufid_out, tfes) {
+                Ok(_) => {}
+                Err(OpteError::MaxCapacity(limit)) => {
+                    return Err(ProcessError::FlowTableFull {
+                        kind: "TCP",
+                        limit,
+                    });
+                }
+                Err(_) => unreachable!(
+                    "Cannot return other errors from FlowTable::add"
+                ),
+            };
+        }
+
+        Ok(tcp_state)
+    }
+
+    /// Attempts to lookup and update TCP flowstate in response to a given
+    /// packet.
+    ///
+    /// Unexpected TCP segments on existing connections will be allowed,
+    /// but will fire DTrace probes via `Self::tcp_err_probe`.
+    ///
+    /// # Errors
+    /// * `ProcessError::MissingFlow` if no flow currently exists.
+    /// * `ProcessError::TcpFlow(NewFlow { .. })` if this packet retired
+    ///   an existing TCP flow state entry.
+    ///
+    /// Callers which expect an existing UFT entry should resubmit a packet
```

Comment:
Could you specify where in the packet processing pipeline such a resubmission should take place? At face value, I can't tell if this is referring to this function or somewhere else.